File: js/ctci/1-arrays-and-strings/1.5-one-away.js
function oneAway(inputStr, outputStr) {
  const hash = {};
  const str1 = inputStr.split('');
  const str2 = outputStr.split('');
  for (let i = 0; i < str1.length; i++) {
    hash[str1[i]] = 1;
  }
  for (let i = 0; i < str2.length; i++) {
    if (hash[str2[i]] === 1) {
      delete hash[str2[i]];
    }
  }

  return Object.keys(hash).length === 1;
}

console.log(oneAway('pale', 'ple')); // true
console.log(oneAway('pales', 'pale')); // true
console.log(oneAway('bale', 'pale')); // true
console.log(oneAway('pale', 'bake')); // true

File: python/practice/ctci/chp01_array_and_strings/1.5-one-away.py
def one_away(s1, s2):
    if (abs(len(s1) - len(s2)) > 1):
        return False

    hash = {}

    for i in range(len(s1)):
        hash[s1[i]] = 1

    for i in range(len(s2)):
        if (s2[i] in hash):
            del hash[s2[i]]
        else:
            hash[s2[i]] = 1

    return True if sum(hash.values()) == 1 else False


print(one_away('ple', 'pale'))  # True
print(one_away('ple', 'paale'))  # False
